lib/db2.js

```diff
@@ -442,80 +443,66 @@ DB2.prototype.update = function(model, where, data, options, cb) {
-      });
-
-      var columns = ParameterizedSQL.join(setValues, ',');
-      sql.merge(columns);
-      sql.merge(')');
-      sql.merge('USING (SELECT * FROM TABLE( VALUES(');
-
-      setValues = [];
-      for (var i = 0, n = fields.names.length; i < n; i++) {
-        setValues.push(new ParameterizedSQL('CAST (' + columnValues[i].sql +
-          ' AS ' +
-          self.buildColumnType(fields.properties[i]) + ')',
-          columnValues[i].params));
-
-        if (i < (n - 1))
-          setValues[i].sql = setValues[i].sql + ',';
-      }
-
-      sql.merge(setValues);
-      sql.merge('))) AS VT(' + fieldNames.join(',') + ')');
-      sql.merge('ON');
-      sql.merge('(MT.\"' + id + '\" = VT.\"' + id + '\")');
-      sql.merge('WHEN NOT MATCHED THEN INSERT (' +
-      fieldNames.join(',') + ')');
-
-      var values = ParameterizedSQL.join(columnValues, ',');
-      values.sql = 'VALUES(' + values.sql + ')';
-      sql.merge(values);
-    } else {
-      sql.merge(self.buildInsertDefaultValues(model, data, options));
-    }
-
-    sql.merge('WHEN MATCHED THEN UPDATE SET');
-
-    setValues = [];
-    for (i = 0, n = fields.names.length; i < n; i++) {
-      if (!fields.properties[i].id) {
-        setValues.push(new ParameterizedSQL(fields.names[i] + '=' +
-          '(' + columnValues[i].sql + ')', columnValues[i].params));
+    var stmt;
+    var tableName = self.tableEscaped(model);
+    var idData = {};
+    var countData;
+    var isInsert = false;
+    var meta = {};
+    var connection;
+    var currentTx;
+
+    process.nextTick(function() {
+      if (options.transaction) {
+        connection = options.transaction.connection;
+      } else {
+        try {
+          connection = Driver.openSync(self.connStr);
+          currentTx = connection.beginTransactionSync();
+        } catch (e) {
+          return callback(Error(e));
+        }
       }
-    }
-
-    sql.merge(ParameterizedSQL.join(setValues, ','));
 
-    self.execute(sql.sql, sql.params, options, function(err, info) {
-      if (!err && info && info.insertId) {
-        data.id = info.insertId;
-      }
-      var meta = {};
-      if (info) {
-        // When using the INSERT ... ON DUPLICATE KEY UPDATE statement,
-        // the returned value is as follows:
-        // 1 for each successful INSERT.
-        // 2 for each successful UPDATE.
-        meta.isNewInstance = (info.affectedRows === 1);
+      idData.id = data.id;
+
+      try {
+        var countStmt = new ParameterizedSQL('SELECT COUNT(*) AS cnt FROM ');
+        countStmt.merge(tableName);
+        countStmt.merge(self.buildWhere(data));
+
+        countData = connection.querySync(countStmt.sql, countStmt.params);
+
+        if (countData[0]['CNT'] > 0) {
+          stmt = self.buildUpdate(model, idData, data);
+        } else {
+          stmt = self.buildInsert(model, data);
+          isInsert = true;
+        }
+
+        connection.querySync(stmt.sql, stmt.params);
```

Comment:
Same as above.  In order to return proper results of multiple parallel calls this must be synchronous.  For example, some of the tests in the test suites execute this function with an array of rows to insert.  The function is called once per row.  If this were async all callers could get to line 484 and get the exact same value for ID as they query the maximum value in the generated column.  If they are sync then they block each other and return the proper values.